modules.cmake

```diff
@@ -43,25 +43,83 @@ if (CMAKE_CXX_COMPILER_ID MATCHES "Clang")
 endif ()
 
 
+
+# Receives latest available C++ standard version
+#
+# Usage:
+#   modules_get_latest_cxx(<variable_name>)
+#   if(<variable_name> GREATER 17)
+#      ...
+#   endif()
+function(modules_get_latest_cxx result)
+  foreach(ver 26 23 20 17 14 11 98)
+    if(CMAKE_CXX_COMPILE_FEATURES MATCHES "cxx_std_${ver}")
+      set(${result} ${ver} PARENT_SCOPE)
+      break()
+    endif()
+  endforeach()
+endfunction()
+
+
+# Checks that compiler has support for C++ modules feature
+#
+# Usage:
+#   modules_is_supported(<variable_name>)
+#   if(<variable_name>)
+#      ...
+#   endif()
+function(modules_is_supported result)
+    set(${result} FALSE PARENT_SCOPE)
+
+    # check standard version
+    modules_get_latest_cxx(latest_standard)
+    if(latest_standard GREATER_EQUAL 20)
+
+      # create simple module file
+      set(temp_filepath "${CMAKE_CURRENT_BINARY_DIR}/module_test.cc")
+      file(WRITE "${temp_filepath}" "module;\nexport module module_test;\nexport void module_test_fun(){}")
+
+      # set needed compiler flags
+      set(compiler_flags "")
+      if (CMAKE_CXX_COMPILER_ID MATCHES "MSVC")
+        set(compiler_flags "/interface")
+      endif()
+
+      # try to build it
+      set(CMAKE_TRY_COMPILE_TARGET_TYPE STATIC_LIBRARY)
+      try_compile(
+        compilation_result "${CMAKE_CURRENT_BINARY_DIR}"
+        SOURCES "${temp_filepath}"
+        COMPILE_DEFINITIONS "${compiler_flags}"
+        CXX_STANDARD ${latest_standard}
+        CXX_STANDARD_REQUIRED ON
+        OUTPUT_VARIABLE output
+      )
+
+      # return result
+      set(${result} ${compilation_result} PARENT_SCOPE)
+    endif()
+endfunction()
+
+
 # Adds a library compiled with C++20 module support.
 # `enabled` is a CMake variables that specifies if modules are enabled.
-# If modules are disabled `add_module_library` falls back to creating a
-# non-modular library.
 #
 # Usage:
-#   add_module_library(<name> [sources...] FALLBACK [sources...] [IF enabled])
```

Comment:
`IF enabled` is needed by {fmt}, please keep it.

Reply:
returned back